### DISCUSSION TITLE
Announcing Sionna v0.17

### DISCUSSION THREAD

**Discussion by jhoydis** [NVlabs/sionna#415] [Announcements]
# Ray tracing
 
- Every scene object has a now a velocity vector which is used to compute per-path Doppler shifts. These are used by `Paths.apply_doppler()`to compute time evolution of channel impulse responses.
- Every scene object has now  a `position` and `orientation` property that can be modified after a scene is loaded.
- A new tutorial notebook [“Mobility in Sionna RT”](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Mobility.html) demonstrates various ways how the new features can be used to simulate mobility.
 
# Misc
 
- Fixes typos in the 5G NR PUSCH Tutorial (#384)
- Fixes typos in the Neural Receiver Tutorial (#406, contribution by @pablosreyero)
- Fixes typos in Tutorial Notebooks 1 and 2 (#355, contribution by @LateNightIceCream)
- Fixes a bug in the `Demapper` for very high SNR (#327, contribution by @japm48)
- Fixes an issue in `sample_positions()` for coverage maps (#376)
- Removes warnings related to casting from `tf.float` to `tf.complex` (#348)
- Fixes a bug in `theta_phi_from_unit_vec()` so that gradients are always well defined

**Comment by fanyimeng0** [comment]
Hello, I am trying to run the example in Mobility in Sionna RT, but I find that I always fail to run it when loading objects in the scene. I have encountered the same problem on both Windows using pip installation and Debian using docker installation. I make sure that my Sionna has been updated to version 0.17.0. How can I solve this problem?
![1714221879504](https://github.com/NVlabs/sionna/assets/81957249/6f498fc1-906f-401b-83b8-090f7b9ad51c)

**Reply by robpegurri** [reply]
I have the exact same issue

**Reply by jhoydis** [reply]
Could you run the code in a a python script so that one can see the error message and understand what is happening?

**Reply by robpegurri** [reply]
This is the only output I get, I don't know how this could be useful:

`2024-04-29 11:50:45.187605: F tensorflow/core/framework/tensor.cc:773] Check failed: IsAligned() ptr = 0x64bb360
Aborted (core dumped)`

I noticed the same thing happens with Google Colab with CPU set as Hardware accelerator. When changed to T4 GPU, the code seems to work fine (with some alerts anyway). My environment is set to not use GPUs too.

**Reply by jhoydis** [reply]
I can reproduce the error on Colab. It works fine on my CPU.

**Reply by jhoydis** [reply]
I have found the bug. It will be corrected in the next release.
In the meantime, you can make the following change [in these lines](https://github.com/NVlabs/sionna/blob/8ad32bc6eca75cfe4003a6a5f035cf4a435ff84e/sionna/rt/scene_object.py#L148C9-L150C70):
```
bbox_min = mi_to_tf_tensor(mi.Point3f(self._mi_shape.bbox().min), rdtype)
bbox_max = mi_to_tf_tensor(mi.Point3f(self._mi_shape.bbox().max), rdtype)
```

**Reply by robpegurri** [reply]
Thanks Jakob! I did update the lines as you suggested.

Again referring to the example, I do confirm the lines are finally working:
```
car_2 = scene.get("car_2")
print("Position: ", car_2.position.numpy())
print("Orientation: ", car_2.orientation.numpy())
```

By the way, the following code block does not:

```
# Move the car 10m along the y-axis
car_2.position += [0, 10, 0]

# And rotate it by 90 degree around the z-axis
car_2.orientation = [np.pi/2, 0, 0]

if colab_compat:
    scene.render(camera="cam", num_samples=512)
    raise ExitCell
scene.preview(); # Do not open a new preview, the preview above will update
```

The output is the following, hope it helps!

```
Traceback (most recent call last):
  File "sionna_mobility_example.py", line 69, in <module>
    car_2.position += [0, 10, 0]
  File "/home/roberto/.local/lib/python3.8/site-packages/sionna/rt/scene_object.py", line 235, in position
    solver_paths.wedges_origin.scatter_nd_update(wedges_ind, wedges_origin)
  File "/home/roberto/.local/lib/python3.8/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 1424, in scatter_nd_update
    gen_state_ops.resource_scatter_nd_update(
  File "/home/roberto/.local/lib/python3.8/site-packages/tensorflow/python/ops/gen_state_ops.py", line 740, in resource_scatter_nd_update
    _ops.raise_from_not_ok_status(e, name)
  File "/home/roberto/.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 6656, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.InvalidArgumentError: {{function_node __wrapped__ResourceScatterNdUpdate_device_/job:localhost/replica:0/task:0/device:CPU:0}} Dimensions [0,1) of indices[shape=[18,1]] = 18 must match dimensions [0,1) of updates[shape=[1,18,3]] = 1 [Op:ResourceScatterNdUpdate] name: 
```

**Reply by jhoydis** [reply]
Indeed, it seems that the fix causes another issue. We'll look into it.

**Reply by jhoydis** [reply]
You can try:
```
bbox_min = tf.cast(self._mi_shape.bbox().min, rdtype)
bbox_max = tf.cast(self._mi_shape.bbox().max, rdtype)
```
My previous code created a dummy batch dimension.

**Reply by robpegurri** [reply]
Everything works perfectly now, thanks!

**Comment by abrxao** [comment]
Hello, i'm very exicited with this new feature and want to integrate Sumo simulation traffics to Sionna RT. I searched some content to do this, but not succesfully. Can i do this integration? If yes, have some documentation to do this?

**Comment by sabusari** [comment]
Hi,
Regarding v0.18, further to the sample parameters in the RIS tutorial, is the implementation frequency-agnostic? Does it scale to mmWave and THz bands with appropriate configuration? Is the RIS also compatible with multiple antennas at the TX and RX?

PS: Asked here because it appears the v0.18 discussion forum is locked.

**Reply by jhoydis** [reply]
Hi @sabusari,

Yes, you can scale it to any frequency. The unit cell spacing is always $\lambda/2$. And, yes, you can have multiple antennas at the TX and RX.

**Reply by sabusari** [reply]
Great, thanks!

**Comment by Abdulla-qa** [comment]
Hello @jhoydis,

I’m relatively new to Sionna-RT, and I’m currently using v1.0 for a project where I simulate radar sensing for intelligent human detection using machine learning.
Right now, I’m in the data generation stage — trying to collect meaningful Channel Impulse Responses (CIRs) that my ML model can later learn from.

My main goal is to make the CIR values different between these cases:

-   a scene without a human, and

-   a scene with a (possibly moving) human. 

In the first place, I Kept human object static, my CIR values remain almost identical— both in amplitude and energy. Look at the figure plz.
<img width="1161" height="1188" alt="image" src="https://github.com/user-attachments/assets/04a3dfb8-6ad7-49b2-9921-a05519b4a360" />

What I’ve Tried  Static Scene Baseline

1- I load my 3D scene (my_scene) and place a transmitter (drone_2_tx) and receiver (drone_2_rx).

2- I run the PathSolver() once per angle and compute the CIR using cir_to_time_channel.

3- This gives consistent and valid results.

Result (static human):
CIR amplitudes and total signal energy are stable, as expected, no real difference between human vs no human.

Then I thought of three directions to tackle it:

    1- Manual Motion Loop
    
    I introduced motion by updating the human’s position before each PathSolver() call:
    
    ```
    for step in range(MOTION_STEPS):
        human.position = [x0, y0 + offset, z0]
        paths = p_solver(scene=my_scene, ...)
    ```
    
    
    Then I averaged CIRs across frames to maintain fixed vector length.
<img width="560" height="751" alt="image" src="https://github.com/user-attachments/assets/a5ca5494-7df1-4838-bc05-fdc1fd05dfd0" />

    
    Result:
      Motion preview visually works, but CIRs look almost identical to the static case —
      very small or negligible difference (attached plots below).
    
   2-Doppler Attempt
    
    I tried using the new velocity and Doppler API:
    
      ```
    human.velocity = [0, 0.2, 0]
      paths = solver(scene=my_scene)
      paths.apply_doppler(duration, sample_rate)
    
    ```
    
    Unfortunately, I get errors or no observable change.
    I suspect I’m misusing the new Doppler feature or missing a required step.

I can’t rely on paths.objects or object IDs — my ML pipeline only uses CIR amplitudes, delays, and energy as input features.
So I need the motion itself (even small vibrations or walking-like movement) to physically affect the CIR values, creating measurable differences that an ML model could generalize from.



Questions:

1. What’s the correct way in Sionna-RT v1.0 to introduce realistic motion (position or velocity) so that it affects the CIR values?

2. Is there a working example of velocity + apply_doppler() in the current version that I could follow?
 
3. Are there solver parameters or timing settings (e.g., sampling rate, duration) that I should tune to make Doppler or motion effects more noticeable?


I really appreciate any advice or clarification you can offer. I’ve been testing different motion strategies for a while and still can’t get the CIRs to meaningfully change.
If there’s a better way to model human motion or Doppler shifts for radar scenarios in v1.0, I’d be very grateful to learn it.

Thanks a lot for your time!

**Reply by merlinND** [reply]
Hello @Abdulla-qa,

Please open a new Discussion to ask your questions, as they are unrelated to the v0.17 announcement: https://github.com/NVlabs/sionna/discussions/new/choose
